### DISCUSSION TITLE
Recommendations for Use without creating models.ApiGateway's

### DISCUSSION THREAD

**Discussion by nitsujri** [StevenMapes/django-aws-api-gateway-websockets#9] [Q&A]
Hi @StevenMapes!

I found your library and it is incredible. I came over from the [Mangum](https://github.com/jordaneremieff/mangum/) side specifically discussing how to use Mangum with AWS [Websockets](https://github.com/jordaneremieff/mangum/discussions/225).

Mangum's solution was never complete and eventually removed. My team just started becoming interested in Django's channels and made my way here. This is an amazing library, thanks so much for bringing it to the community.

At the moment, it looks like django-aws-api-gateway-websockets (**DAAGW** for short?) requires creating+manging the Websocket gateway via itself. I was wondering if there was any way to allow non-locally created routes.

All of our infra is managed via CDK and so having this one portion managed here makes it significantly less than ideal.

Ideally, if possible, the app stores this information as the connection request comes in, post auth if necessary.

Thanks!
Justin

**Comment by StevenMapes** [comment]
Hi @nitsujri and thank you for the kind comments.

I designed this project with the idea that it would be used within projects where you wanted to use Django admin to create most of the setup within AWS, especially the creation of the API Gateway integration.  From what you're describing you'd like to be able to create the API Gateway setup outside of the project and then have it accept the request as it comes in and then handle it as if it were already setup.

I'll have to have a think about that, currently it would fail whilst calling the self._allowed_apigateway() method as that's designed to only allow expected API Gateways. So one option may be to extend subclass WebSocketView and change the handling of that method to perform the setup. I've not investigated it but it's just a thought at this stage whilst I'm trying to get a better idea of your requirements as I've not used Mangum myself.

One way other way would be if I introduced DRF support to expose API endpoints to the tables. Then you could populate the tables via API and call call endpoints to trigger the createApiGateway and createCustomDomain management commands but I don't think this will quite do what you want of dynamically setting it up as part of the request.

**Reply by nitsujri** [reply]
Thanks for the response!

After reading the library more deeply, I'll preface with - I don't think our use case is a good reason to modify your library. The philosophies are quite different and I think your beliefs should stay.

Our org believes in infra as code, hence the CDK and no clicking on AWS console for prod/uat, which I suppose extends to here as well.

With that our infra looks like:

1. Build container -> ECR
2. CodeDeploy -> [aws_sam.CfnFunction](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda.CfnFunction.html)
   - There's an AWS Lambda for Website, Worker and Websocket. 3 diff lambdas using the same Container & diff entrypoints.
3. Websocket Lambda -> API Gateway as [WebSocketLambdaIntegration](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_apigatewayv2_integrations.WebSocketLambdaIntegration.html)
   - Using `WebSocketLambdaIntegration` means we don't have direct control of the `integration_url`. So, if we wanted more routes, we'd have to build separate Lambdas w/ entrypoints. Realistically, this means everything goes through $default.
4. [API Gateway.url and callbackUrl](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_apigatewayv2.WebSocketStage.html#properties) -> Website Lambda as environment vars.
   - This is how our app knows what the `wss://` is

All this to say our app doesn't "know or control" infra. It gets told how to interface with it.

For us, I think you're right. Looking at `self._allowed_apigateway` and `self._check_platform_registered_api_gateways` to override as an entry points to create `ApiGateway`+`ApiGatewayAdditionalRoute`s.

Possibly, I think a more interesting Feature request for this library is an agnostic API layer, such that Django Channels or AWS WebsocketsAPI can be used seamlessly.

Should I write a Feature Request for this?

**Reply by StevenMapes** [reply]
So I've create a feature request to enable self registering gateways. It occurred to me that it would be good feature to have and would be secure as it'll be controlled by the IAM being used at the time so I've created #10 as a request for that. My plan is that if the ApiGateway does not exist within the datbase that the code then pulls all the APIGateways via boto3 and sets them up so that they should be available.  That way the infrastructure could be setup outside of the project.

Would that match your use case? If not then feel free to create feature requests with what you're after and I'll have a think about them. I'm not sure I fully understand why you'd use this with Channels though as usually if you're using channels it's controlling the websockets so you wouldn't use API Gateway

**Reply by nitsujri** [reply]
> Would that match your use case?

Yes, and a few thoughts:

- The boto3 call would be slow on first hit. 
- Possibly the boto3 is unnecessary because as long as the `HTTP/Lambda proxy integration` is enabled, `domainName` `stage` and `connectionId` available in the `requestContext`. That's enough to `post_to_connection`.
  - Admittedly negates a lot of the required fields of `ApiGateway` and all `ApiGatewayAdditionalRoute`
  - Unsure, the proxy may not be required if the data is carried in `context` param, but I haven't looked.

> I'm not sure I fully understand why you'd use this with Channels though as usually if you're using channels it's controlling the websockets so you wouldn't use API Gateway

For us the reason is local dev vs server based testing.

Thanks for the discussions, it's quite enjoyable.

**Reply by StevenMapes** [reply]
Hmm interesting. My thoughts on using boto3 was more around validating that the API Gateway making the request was from an AWS account you had access to, so effectively acting as a whitelist. So my plan was to validate that the IAM had access as otherwise it is effectively a public endpoint that anyone could use to send messages in. I would need to do some further investigation based on your comments

**Reply by nitsujri** [reply]
> validating that the API Gateway making the request was from an AWS account you had access to, so effectively acting as a whitelist

If it's possible, that's infra security and ideally be secure by default. App servers on private VPCs; All external requests through API Gateway, CloudFront+WAL; access via secgrps...

But that's the ideal and many many apps don't run that setup because it's a headache with a steep learning curve. Like Lambdas on assigned a VPC can't access S3 directly anymore when they could.

So I think your belief is very valid and checking+ensuring ownership is good.

**Reply by StevenMapes** [reply]
Okay that in mind that's what I could do is to set class attributes that disabled the need to check which could be used in situations where the infrastructure has been set up to take care of the security such as private vpcs etc. I will update my feature request to reference this as then it could be turned on for projects where this app doesn't need to worry about the security. So this is being a very useful discussion as it is giving me more ideas for features to help improve everything.

**Reply by StevenMapes** [reply]
I question for you @nitsujri. I've started working on this feature but I need the stage name in order to send messages. I've edited one of my existing API Gateways to turn on the HTTP Proxy flag and have redeployed it but I've not noticed and change in the headers that come in so I'm still not being passed the route. The only headers I'm receiving are:

'Host': 
'X-Real-Ip':
'X-Forwarded-For': 
'X-Forwarded-Proto':
'Connection':
'Content-Length':
'X-Forwarded-Port': 
'X-Amzn-Trace-Id': 
'Connectionid':
'X-Api-Key':
'X-Restapi':
'User-Agent':
'X-Amzn-Apigateway-Api-Id':

In your experience are you receiving different headers when you have API Gateway sending to Lamda?

For the implementation where I'm going to validate self-registering gateways against the AWS account I will be able to look up the stage name but for the case where it's "Infra secured" and I want to allow self-registering without checking we'll need to get the stage. Currently this app will default to "production".

It does remind me that I need to create a feature request for supporting multiple stages per API Gateway as currently I do not support that.
